enhancement(networking): Automatic request limiting

bruceg opened the pull request:
This is my work so far on the automatic request limiting framework. The new `AutoConcurrencyLimit` struct should be a drop-in replacement for `ConcurrencyLimit`. It is not usable as it is, due to a trait issue.

I have introduced a new trait `IsBackPressure` that is required on the error type to indicate if it is a back-pressure error or some other error. However, the base trait used in services is `Timeout<S>`: https://github.com/timberio/vector/blob/b56f7374edf4e61d3ac9609df6715c197cdd12d5/src/sinks/util/service2.rs#L17
[`Timeout` defines its `Error` type as `Box<dyn Error + …>`](https://docs.rs/tower/0.3.1/tower/timeout/struct.Timeout.html#impl-Service%3CRequest%3E) which can't be made to `impl IsBackPressure` (can it?). I'm stumped at the moment on how to resolve this.

Closes #2529

Commit:
Import base concurrency limiter from tower

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop unused Load impl on AutoConcurrencyLimit

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce Controller wrapper for the semaphore

The implementation is complicated by the fact that we can only decrease
concurrency by acquiring permits and forgetting them. If there aren't
enough free permits, this requires waiting for them when polling for
readiness.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Adjust pub markers and drop more unused methods

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Link the semaphore controller and start time into the response future

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Calculate (and throw away) RTT in response future

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add function to controller to calculate average RTT

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add increase and decrease methods to the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Initial implementation of RTT-based concurrency adjustment

Has issues that need correcting:

* has recursive locking via `fn contract` and `fn expand`
* adjusts on every measurement, not once per RTT
* adjusts for any response, does not differentiate

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix recursive locking when making adjustments, and rename "dropping"

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Limit adjustments to once per RTT

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce new EWMA abstraction

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Aggregate responses for each RTT interval

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Simplify retry logic with `matches!`

Signed-off-by: Bruce Guenter <bruce@timber.io>

LucioFranco reviewed (commented):
looks like this is heading in the right direction! Obviously, I'd like to see some tests but overall looks great!

bruceg commented on Cargo.toml:
```diff
@@ -42,8 +42,9 @@ tracing-limit = { path = "lib/tracing-limit" }
 # Tokio / Futures
 futures01 = { package = "futures", version = "0.1.25" }
 futures = { version = "0.3", default-features = false, features = ["compat"] }
+futures-core = "0.3"
```
Got it. The code I copied from imported from `futures_core` which required that import.

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,189 @@
+        MaybeForgetFuture {
+            semaphore: self.semaphore.clone(),
+            inner: self.inner.clone(),
+            future: Box::pin(Arc::clone(&self.semaphore).acquire_owned()),
```
Because without the pin, I get method not found on `future.poll`. How does the status of `OwnedPermit` relate to this future not needing to be pinned? `acquire_owned` returns a future, not the permit directly, since it is an `async fn`.

LucioFranco replied:
Ah you're right! since it is a async fn its always going to be `!Unpin` 👍 good catch

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,189 @@
+        self._adjust_to_response(start, response)
+    }
+
+    fn _adjust_to_response(&self, start: Instant, response: ResponseType) {
```
The non-underscore is just a wrapper function for the inner one, parrtly to avoid monomorphization of a large function. This pair evolved over a series of tries. If I move to a fixed type in the future, this can go away.

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,189 @@
+    {
+        let response = match response {
+            Ok(_) => ResponseType::Normal,
+            Err(r) if r.is_back_pressure() => ResponseType::BackPressure,
```
An alternative, as you suggest, may be to extend the retry logic type to include the needed functionality (if it doesn't already) and use that instead of a new type. It might still need additional trait object, but at least there wouldn't be the need to implement another trait.

lukesteensen reviewed (commented):
This is looking really good! I agree it'd be good to exercise this pretty thoroughly with tests, but the overall shape is solid.

I'm also wondering if it'd be helpful to more closely wrap `semaphore`, `current`, `max`, and `to_forget`. There's some complex overlapping state there, and having a single API that adjusts them all together could help avoid consistency bugs.

Commit:
Wrap the shrinkable semaphore in its own module

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
> I'm also wondering if it'd be helpful to more closely wrap `semaphore`, `current`, `max`, and `to_forget`. There's some complex overlapping state there, and having a single API that adjusts them all together could help avoid consistency bugs.

I wrapped up the semaphore bits (at the expense of introducing another `Arc` layer), does that help? Do you see more opportunities for introduction of wrappers?

lukesteensen reviewed (commented):
That looks good! I'm only pro-wrapper where they make it simpler to maintain invariants, so I think that's probably enough :smile:

Commit:
Use new stub AutoConcurrencyLimit layer in service2-based sinks

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add RetryLogic to auto concurrency layer (unused)

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Handle back pressure through applying RetryLogic (BROKEN)

I think the problem is that the Timeout layer needs to be inside the
AutoConcurrencyLimit layer, but the Error type in the Timeout layer is a
Box<Error> which differs from the RetryLogic Error type.

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented:
@Hoverbear adding this to your sprint to help unblock @bruceg on the error mapping issue.

Hoverbear committed:
Fixup ;)

Signed-off-by: Ana Hobden <operator@hoverbear.org>

Commit:
Start work on tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix: Make sure there is an initial set of measurements

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix: Only reset the current measurements after using them

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Start of auto request limiting test framework

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Silence a clippy warning in tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move `get_controller` and `capture_metrics` into src/metrics.rs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move creation of Metric from a Measurement into `src/event/metric.rs`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce an `assert_between` macro for range matches

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename `current` to `current_limit` in controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add and test internal metrics for the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move autoreq test parameters into their own struct

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Generate the test statistics all in one go, and add mode

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename assert_between to assert_within

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix handling of error responses in the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add tests for dropped and deferred requests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add get_ref methods to access internals of our service layers

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Make test_util into a module

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move autoreq tests into the src tree beside auto_concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Start of internal controller statistics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Update Cargo for new get_ref methods in tokio03

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop unused use std::error::Error

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
Replaced by #3094
